### ISSUE TITLE
Documentation request: Finding the root of the current workspace

### ISSUE THREAD

**Issue by cpitclaudel** [ocaml/dune#2856]
Hi there,

Thanks for the hard work on dune.

**TL;DR:** should `%{project_root}` and `%{workspace_root}` expand to absolute paths?

Context: I'm trying to generate part of a dune file; the documentation suggests this: `(rule (with-stdout-to dune.inc.gen (run ./gen-dune.exe)))`.

Assume `gen-dune.exe` is compiled from `gen_dune.ml`, and is used to generate one rule per `.xyz` file in the project's root.

`gen_dune.ml` has to enumerate `.xyz` files in the project's root, but it's not trivial for it to know where that is, since its run from the build directory.  Changing the rule to `(rule (with-stdout-to dune.inc.gen (run ./gen-dune.exe %{project_root})))` runs `gen-dune.exe` with argument `.`, which is the path from `gen_dune.ml`, not from `gen-dune.exe`, nor an absolute path.

Is there a way to get an absolute path? Am I doing something wrong?

Thanks!

**Comment by cpitclaudel** [comment]
Note: I saw the note about `chdir`, but `(with-stdout-to dune.inc.gen (chdir %{project_root} (run ./gen_dune.exe)))` says `No rule found for gen_dune.exe`.

**Comment by diml** [comment]
That seems odd, in `(rule (with-stdout-to dune.inc.gen (run ./gen-dune.exe %{project_root})))` `%{project_root}` should definitely expand to something like `../..`, unless this `dune` is at the root already.

Regarding the second error, you can replace `./gen_dune.exe` by `%{exe:gen_dune.exe}`. The argument of `%{exe:...}` is always interpreted relative to the directory where the `dune` file appears, but it expands to a path relative to the runtime directory. So `(chdir %{project_root} (run %{exe:foo.exe}))` usually expands to something of the form `(chdir ../.. (run src/blah/foo.exe))`.

Regarding using absolute paths, it's usually not great to use absolute paths as it makes things less shareable between workspaces. In particular, we are currently working on adding a shared cache to Dune which means that we'll share artifacts between workspaces.

Finally, `dune` should know that `gen-dune.exe` will list all the `.xyz` files. Have you considered the following possibility:

```scheme
(rule
 (deps (:xyz-files (glob_files %{project_root}/*.xyz)))
 (action (with-stdout-to dune.inc.gen (run ./gen_dune.exe %{xyz-files}))))
```

?

**Comment by cpitclaudel** [comment]
> That seems odd, in (rule (with-stdout-to dune.inc.gen (run ./gen-dune.exe %{project_root}))) %{project_root} should definitely expand to something like ../.., unless this dune is at the root already.

OK, that makes sense.  I believe this is a bug, then.

> Regarding the second error, you can replace ./gen_dune.exe by %{exe:gen_dune.exe}

Brilliant, thanks!

> Regarding using absolute paths, it's usually not great to use absolute paths as it makes things less shareable between workspaces.

Got it.

> Finally, dune should know that gen-dune.exe will list all the .xyz files. Have you considered the following possibility.

Thanks a good suggestion, thanks.

**Comment by diml** [comment]
I couldn't reproduce the bug BTW. Do you have a reproduction case?

**Comment by cpitclaudel** [comment]
Sorry, it seems I dropped the ball on that issue.  Not sure why it fell off my radar.  Sincere apologies :/

I've returned to this issue of generating dune files.  I think your suggestion is great (re `*.deps`).
But I still have one problem.  Consider the following rule, which for simplicity handles just one file


```
(rule
 (mode (promote (until-clean)))
 (deps
  (:xyz example.xyz))
 (targets
   xyz.cpp
   xyz.dot
   xyz.hpp
   xyz.kpkg
   xyz.verilator.cpp
   cuttlesim.hpp
   Makefile
   verilator.hpp)
 (action
  (chdir %{{workspace_root}}
   (run cuttlec %{{xyz}} -T all -o .))))
```

We have a compiler that takes xyz files and makes a bunch of outputs, that it puts together in a directory indicated by the `-o` option.  The chdir ensures that (as recommended in the docs) we get error messages relative to `dune-project`.  But then there's a problem: the `.` isn't quite right.  It should be the target directory — only I haven't found a way in dune to talk about this directory.

It would be great to have a function of some sort to indicate that something is a path and should be adjusted relative to the current working directory.  This way I'd write `%{relpath:.}`, or something like this.  Should I open a separate issue about this?

Thanks again.

**Comment by jeremiedimino** [comment]
Adding such a form seems reasonable to me. Would you be happy to write the PR for it?

**Comment by rgrinberg** [comment]
@gridbugs what do you think about `%{relpath:.}` in the comment above?

**Comment by gridbugs** [comment]
Slightly tangential to your question but I think there's a bigger issue that this connects to:

I've been thinking a bit about paths in dune lately after encountering https://github.com/ocaml/dune/issues/7146. A pain point for dune at the moment is that some commands are run from the current directory (e.g. user actions) and other commands are run from the workspace root (e.g. the compiler). There is nothing inherently wrong with running some commands from the workspace root _except_ that we let the user specify custom arguments to these commends in dune files (e.g. `c_library_flags`), and some of those arguments can contain paths. In my experience most of the time when a path appears in a dune file its interpretation is intended to be relative to the dune file's directory or from the project's root directory, so users are trained to use relative paths or paths beginning with `%{project_root}` (which itself is relative to the dune file's directory), and it's an unpleasant surprise when a command run from outside the current directory fails due to being given a relative path, or worse, _doesn't_ fail and operates on the wrong file due to a relative path being resolved in an unexpected place (here's an example of this happening in the wild which was benign except for breaking vendoring https://github.com/c-cube/ocaml-qbf/pull/14).

Furthermore we have the `chdir` combinator, but it's tricky to use paths in a `chdir`'d command, especially because `%{project_root}` is a path relative to the dune file's directory.

The `%{relpath:...}` proposal seems like it could address the `chdir` issue as well as the issue of some commands running outside the current directory. I assume it would take a path relative to the current directory and return the equivalent path relative to wherever the command ends up being run from? In addition to using it in `chdir`, I imagine doing something like `(c_library_flags -L%{relpath:.} ...)` to let the linker find libraries in the current directory despite it being run from the workspace root.

However I find the fact that `%{relpath:...}` would behave differently depending on the context where it is used a little daunting. My preference would be to instead introduce `%{abspath:...}` which would take a path relative to the current directory and return the absolute path to that file. I think this would also solve both the `chdir` and `c_library_flags` problems but would behave consistently regardless of where it appears in a dune file.

Also, my hot take about `%{project_root}` is that I think it's dangerous to give users a `%{project_root}` variable if it's only valid for it to appear in certain contexts within a dune file (ie. arguments to commands that are run from the dune file's directory). I would be interested in (re?)opening the discussion around making `%{project_root}` an absolute path to partially address the above. It's confusing that `%{project_root}` can be used in some places but not others. Up the conversation I see:
>  Regarding using absolute paths, it's usually not great to use absolute paths as it makes things less shareable between workspaces.

I can't think of a situation where making `%{project_root}` an absolute path would make things less shareable. I see that the person who posted this has deactivated their account, but I would appreciate an example of a problem that could arise from switching to absolute paths.

**Comment by nojb** [comment]
> I would appreciate an example of a problem that could arise from switching to absolute paths.

A big downside would be that it would no longer be possible to pass `%{project_root}` to programs expecting Unix-style paths.

**Comment by gridbugs** [comment]
What do you mean by unix-style path?

**Comment by nojb** [comment]
> What do you mean by unix-style path?

A path that makes sense to a program in e.g. Cygwin (in particular, not referring to the Windows root hierarchy `C:\`, using forward slashes, etc).

**Comment by gridbugs** [comment]
Ah I see. In non-unix systems is it ever the case that some programs expect unix-style paths and some do not?

**Comment by nojb** [comment]
> Ah I see. In non-unix systems is it ever the case that some programs expect unix-style paths and some do not?

Yes, native Windows programs will typically expect Windows paths (backslashes, referring to the "true" filesystem hierarchy), while Cygwin programs will expect paths that follow Cygwin conventions (forward slashes, referring to the Cygwin filesystem hierarchy, etc).

**Comment by gridbugs** [comment]
Do the native windows programs understand relative paths with forwards slashes? Like can they handle windows-style relative paths with forward slashes and unix-style relative paths with backslashes?

**Comment by nojb** [comment]
> Do the native windows programs understand relative paths with forwards slashes? Like can they handle windows-style relative paths with backslashes and unix-style relative paths with forward slashes?

Indeed, in many (but not all) contexts forward slashes can be passed to native Windows programs.
